### PR TITLE
Improve after resume failure logic

### PR DESCRIPTION
**Pull request by ikbalkaya**
This PR fixes two issues

* When `queueEvents` is false, pending messages will also be cleared which was missing before. 
* Fixes the issue of reattachment when a connection resume failed.

Previously as soon as connected message was received by connection manager, channels would directly attempt to attach. I have changed the behaviour so that after a resume failure (connected event with a new connection id) I transfer all queued messages (if any) from the connection manager to associated channel. This sets a flag for later to be processed from `setConnected`. Previously when attempted to call directly to attach from reattach, that would also fail and retry would happen quite a bit later. Attach would time out and then a reattach with some delay would be invoked. I also removed a sync() call from `setConnected()` as recommended by @paddybyers . 

There was also an issue with previously enters members wouldn't automatically reenter. There should be a fix for this and a test that confirms the behaviour.

There are also some test and test utilities improvements.

Fixes https://github.com/ably/ably-java/issues/905
Fixes https://github.com/ably/ably-java/issues/904

### CODE REVIEW

File: lib/src/main/java/io/ably/lib/realtime/ChannelBase.java [hunk 1]
```diff
@@ -186,6 +186,36 @@ void attach(boolean forceReattach, CompletionListener listener) {
...
+        for (QueuedMessage queuedMessage : messagesToTransfer) {
+            if (queuedMessage.msg.action == Action.message) {
+                queuedMessages.add(queuedMessage);
+            }else if (queuedMessage.msg.action == Action.presence) {
+                PresenceMessage[] presenceMessages = queuedMessage.msg.presence;
+                if (presenceMessages != null && presenceMessages.length > 0){
+                    for (PresenceMessage presenceMessage : presenceMessages) {
+                        String clientId;
+                        try {
+                            clientId = ably.auth.checkClientId(presenceMessage, false, true);
```
Comment (ikbalkaya) [inline lib/src/main/java/io/ably/lib/realtime/ChannelBase.java:203]: I took this part from an existing code which puts pending presence.

File: lib/src/main/java/io/ably/lib/realtime/ChannelBase.java [hunk 2]
```diff
@@ -213,13 +214,13 @@ void reattach(List<QueuedMessage> messagesToTransfer) {
 
             }
         }
-        attach(true,null);
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/realtime/ChannelBase.java]: why remove this?

File: lib/src/main/java/io/ably/lib/realtime/AblyRealtime.java [hunk 3]
```diff
@@ -236,10 +236,11 @@ public void suspendAll(ErrorInfo error, boolean notifyStateChange) {
 
         /**
          * By spec RTN15c3
+         * Move queued messages from connection manager to their respective channel
          * @param queuedMessages Queued messages transferred from connection
          */
         @Override
-        public void reattachOnResumeFailure(List<ConnectionManager.QueuedMessage> queuedMessages) {
+        public void transferQueuedMessages(List<ConnectionManager.QueuedMessage> queuedMessages) {
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/realtime/AblyRealtime.java:243]: As written, this is very inefficient. We're iterating through all queued messages multiple times, once for each channel.

Comment (ikbalkaya) [reply lib/src/main/java/io/ably/lib/realtime/AblyRealtime.java:244]: aa1bc86296f22e21386630394f75f81ae22675c0

File: lib/src/main/java/io/ably/lib/realtime/ChannelBase.java (hunk 2 above)
Comment (ikbalkaya) [reply lib/src/main/java/io/ably/lib/realtime/ChannelBase.java]: I think when running tests I had a wrong impression that this reattach will happen as a result of another state transition as I was seeing channels reattached. but now I see they are stuck in attaching state. It's a result of writing things in hurry without having time to check their effect. I'll reintroduce this.

File: lib/src/main/java/io/ably/lib/realtime/ChannelBase.java [hunk 4]
```diff
@@ -186,6 +193,40 @@ void attach(boolean forceReattach, CompletionListener listener) {
...
+    /**
+     * This method carries queued messages accumulated on connection manager while the channel
+     * isn't attached yet. It's added in the queue here and start a new attach call
+     * */
+    void reattach(List<QueuedMessage> messagesToTransfer) {
+        state = ChannelState.attaching;
+        if (messagesToTransfer != null) {
+            for (QueuedMessage queuedMessage : messagesToTransfer) {
+                if (queuedMessage.msg.action == Action.message) {
+
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/realtime/ChannelBase.java:205]: whitespace

File: lib/src/main/java/io/ably/lib/realtime/ChannelBase.java [hunk 5]
```diff
@@ -186,6 +193,40 @@ void attach(boolean forceReattach, CompletionListener listener) {
...
+                    queuedMessages.add(queuedMessage);
+                }else if (queuedMessage.msg.action == Action.presence) {
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/realtime/ChannelBase.java:207]: please no `}else` - make whitespace consistent

File: lib/src/main/java/io/ably/lib/realtime/ChannelBase.java [hunk 6]
```diff
@@ -186,6 +193,40 @@ void attach(boolean forceReattach, CompletionListener listener) {
...
+                    PresenceMessage[] presenceMessages = queuedMessage.msg.presence;
+                    if (presenceMessages != null && presenceMessages.length > 0){
+                        for (PresenceMessage presenceMessage : presenceMessages) {
+                            String clientId;
+                            try {
+                                clientId = ably.auth.checkClientId(presenceMessage, false, true);
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/realtime/ChannelBase.java:213]: why is this being checked here? If the message had previously been queued, don't we know that the `clientId` is valid?

File: lib/src/main/java/io/ably/lib/realtime/ChannelBase.java [hunk 7]
```diff
@@ -84,6 +85,12 @@ public abstract class ChannelBase extends EventEmitter<ChannelEvent, ChannelStat
 
     private int retryCount = 0;
 
+    /**
+     * This flag is determining whether the attach should be reattampted when a setConnected() event us receuved
+     * This is a workaround and the mechanism to handle this should be rethought in the future
+     * */
+    private final AtomicBoolean reattachOnConnectionResume = new AtomicBoolean(false);
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/realtime/ChannelBase.java:92]: Can you explain what contention this is addressing?

File: lib/src/main/java/io/ably/lib/realtime/ChannelBase.java [hunk 8]
```diff
@@ -406,6 +447,7 @@ synchronized private void clearAttachTimers() {
                 t.cancel();
                 t.purge();
             }
+
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/realtime/ChannelBase.java:450]: why a blank line?

File: lib/src/main/java/io/ably/lib/realtime/ChannelBase.java [hunk 9]
```diff
@@ -624,7 +663,7 @@ public synchronized void setSuspended(ErrorInfo reason, boolean notifyStateChang
             Log.v(TAG, "setSuspended(); channel = " + name);
             presence.setSuspended(reason);
             setState(ChannelState.suspended, reason, false, notifyStateChange);
-            failQueuedMessages(reason);
+            // failQueuedMessages(reason);
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/realtime/ChannelBase.java:666]: why is this removed?

File: lib/src/main/java/io/ably/lib/transport/ConnectionManager.java [hunk 10]
```diff
@@ -1252,19 +1255,16 @@ private synchronized void onConnected(ProtocolMessage message) {
     private void addPendingMessagesToQueuedMessages(boolean resetMessageSerial) {
         // Add messages from pending messages to front of queuedMessages in order to retry them
         queuedMessages.addAll(0, pendingMessages.queue);
-        //rewind start serial back to the first serial since we are clearing the queue
-        if (!pendingMessages.queue.isEmpty()){
-            //Reset current serial to the first pending message on previous queue as we are going to clear the queue now
+        //rewind start serial back to the first serial since we are clearing the queue if the queue is not empty
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/transport/ConnectionManager.java:1258]: I think this would be a lot clearer if it was:

```
if (resetMessageSerial){  // failed resume, so all new published messages start with msgSerial = 0
  msgSerial = 0; //msgSerial will increase in sendImpl when messages are sent
  pendingMessages.resetStartSerial(0);
} else if(!pendingMessages.queue.isEmpty()) { // pendingMessages needs to expect next msgSerial to be the earliest previously unacknowledged message
  msgSerial =  pendingMessages.queue.get(0).msg.msgSerial;
  pendingMessages.resetStartSerial((int) (msgSerial));
}

```

File: lib/src/test/java/io/ably/lib/test/realtime/RealtimeResumeTest.java [hunk 11]
```diff
@@ -909,7 +924,14 @@ public void onConnectionStateChanged(ConnectionStateChange state) {
             assertNotNull(ably.connection.id);
             assertNotEquals("Connection has the same id", firstConnectionId, ably.connection.id);
 
-            /* wait for the publish callback to be called.*/
+            // wait for channel to get attached
+            (new ChannelWaiter(senderChannel)).waitFor(ChannelState.attached);
+            assertEquals("Connection has the same id", ChannelState.attached, senderChannel.state);
+
+            try {
```
Comment (paddybyers) [inline lib/src/test/java/io/ably/lib/test/realtime/RealtimeResumeTest.java:931]: why do we need this wait?

**Review by paddybyers — COMMENTED** [review]
See inline comments

File: lib/src/main/java/io/ably/lib/realtime/ChannelBase.java (hunk 7 above)
Comment (ikbalkaya) [reply lib/src/main/java/io/ably/lib/realtime/ChannelBase.java:92]: `setConnected()` and `reattach()` calls are done from different threads. setConnected() is called when a state transition happens - so it technically will always be done after `setConnected` , but as they arrive from different threads I wanted to minimize any sync issues by using an atomic boolean

File: lib/src/main/java/io/ably/lib/realtime/ChannelBase.java (hunk 5 above)
Comment (ikbalkaya) [reply lib/src/main/java/io/ably/lib/realtime/ChannelBase.java]: c4f34de7056b6c90eeaf8a81a36bdb586bb6b157

File: lib/src/main/java/io/ably/lib/realtime/ChannelBase.java (hunk 4 above)
Comment (ikbalkaya) [reply lib/src/main/java/io/ably/lib/realtime/ChannelBase.java]: c4f34de7056b6c90eeaf8a81a36bdb586bb6b157

File: lib/src/main/java/io/ably/lib/realtime/ChannelBase.java (hunk 8 above)
Comment (ikbalkaya) [reply lib/src/main/java/io/ably/lib/realtime/ChannelBase.java]: c4f34de7056b6c90eeaf8a81a36bdb586bb6b157

File: lib/src/main/java/io/ably/lib/realtime/ChannelBase.java (hunk 7 above)
Comment (paddybyers) [reply lib/src/main/java/io/ably/lib/realtime/ChannelBase.java:92]: which threads? What triggers `reattach()` other than connection state changes?

My concern is that we're mixing synchronisation primitives; most operations involving connection state are performed by the `ConnectionManager` thread, and there is a lock there that protects the `ConnectionManager` state and ensures that all state transitions occur strictly serially.

File: lib/src/main/java/io/ably/lib/transport/ConnectionManager.java (hunk 10 above)
Comment (ikbalkaya) [reply lib/src/main/java/io/ably/lib/transport/ConnectionManager.java]: 57c9a142260dd198c40f054787a7aea6cb0d9542

File: lib/src/main/java/io/ably/lib/realtime/ChannelBase.java (hunk 7 above)
Comment (ikbalkaya) [reply lib/src/main/java/io/ably/lib/realtime/ChannelBase.java:92]: Here is a quick run to check in which thread this value is set and read.
```
2023-01-30 11:11:23.844  8588-8629  System.out              io.ably.lib.test                     I  thread_test reattachOnConnectionResume is set :WebSocketConnectReadThread-341
2023-01-30 11:11:23.845  8588-8617  System.out              io.ably.lib.test                     I  thread_test reattachOnConnectionResume is read :Thread-2
```
I think I misrenamed `reattach` as it isn't technically a reattach and will just set this flag so that this could be used when a state transition happens. I'm equally happy to make this non-atomic as I think operations still be ordered as you mentioned above.

File: lib/src/main/java/io/ably/lib/realtime/ChannelBase.java (hunk 9 above)
Comment (ikbalkaya) [reply lib/src/main/java/io/ably/lib/realtime/ChannelBase.java]: Sorry - this was a result of me furiously commenting out and trying different things and forgetting when cleaning up the code. I believe I commented this out when the channels would go into suspended state as attach wouldn't happen. I put it back f525dcac9fcaef94d1b7893860adb125d9eb307f

File: lib/src/main/java/io/ably/lib/realtime/ChannelBase.java (hunk 6 above)
Comment (ikbalkaya) [reply lib/src/main/java/io/ably/lib/realtime/ChannelBase.java]: I took this from an existing code 
https://github.com/ably/ably-java/blob/1c41924f08dffb1e94adc4ee730584b14f16103e/lib/src/main/java/io/ably/lib/realtime/Presence.java#L743
As I did not fully know the condition of clientId being valid here, I thought this check might be sensible. But I can understand why it's not needed here.
Change: c41824dc911210265ea8f3b55c998206f74ce953

File: lib/src/main/java/io/ably/lib/transport/ConnectionManager.java [hunk 12]
```diff
@@ -1201,18 +1215,22 @@ private synchronized void onConnected(ProtocolMessage message) {
...
-                channels.reattachOnResumeFailure();
-                // Add any messages still pending from the previous transport (RTN19a) to the front of queued messages
-                // however, this time the pending messages have to have newly assigned `
-                // msgSerial`s. They can't simply be replayed, as they are in the successful resume case
+                //we are going to add pending messages and update pending queue state
                 addPendingMessagesToQueuedMessages(true);
+
+                //We are going to transfer those messages to channel level so that they are published after
+                //their respective channels are attached
+                channels.transferToChannels(new ArrayList<>(queuedMessages));
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/transport/ConnectionManager.java:1231]: I think we're going to have to change how we process unacknowledged `Message`s (as distinct from `PresenceMessage`s. According to the spec (https://sdk.ably.com/builds/ably/specification/main/features/#RTL6c2) all queued `Message`s (ie those that were queued to be sent, but not yet sent, in `ConnectionManager.queuedMessages`; plus those that were sent but are unacknowledged, in `ConnectionManager.pendingMessages`) should be re-sent immediately on establishment of the new connection.

**Review by paddybyers — CHANGES_REQUESTED** [review]

File: lib/src/main/java/io/ably/lib/realtime/ChannelBase.java (hunk 7 above)
Comment (ikbalkaya) [reply lib/src/main/java/io/ably/lib/realtime/ChannelBase.java]: As discussed this is no longer the case as this information will be carried out with `StateIndication` c63183dbf8cf68ad1d5d0b60508205779f2defba

**Review by paddybyers — APPROVED** [review]
LGTM, thanks
